File: frontend/src/backend/patientService.js
import {config, appFetch, setServiceToken, getServiceToken, removeServiceToken, setReauthenticationCallback} from './appFetch';
import { getCountries, getAlcoholicDrinker, getBowelFunction,getGenres,getMaritalStatus,getPhysicalActivity,
getSleepQuality,getSmoker } from '../modules/app/actions';

/*export const getGenres = (onSuccess, onErrors) =>
    appFetch('/patients/getGenres', config('GET'), onSuccess, onErrors);

export const getCountries = (onSuccess, onErrors) =>
    appFetch('/patients/getCountries', config('GET'), onSuccess, onErrors);

export const getMaritalStatus = (onSuccess, onErrors) =>
    appFetch('/patients/getMaritalStatus', config('GET'), onSuccess, onErrors);

export const getBowelFunction = (onSuccess, onErrors) =>
    appFetch('/patients/getBowelFunction', config('GET'), onSuccess, onErrors);

export const getSleepQuality = (onSuccess, onErrors) =>
    appFetch('/patients/getSleepQuality', config('GET'), onSuccess, onErrors);

export const getPhysicalActivity = (onSuccess, onErrors) =>
    appFetch('/patients/getPhysicalActivity', config('GET'), onSuccess, onErrors);

export const getSmoker = (onSuccess, onErrors) =>
    appFetch('/patients/getSmoker', config('GET'), onSuccess, onErrors);

export const getAlcoholicDrinker = (onSuccess, onErrors) =>
    appFetch('/patients/getAlcoholicDrinker', config('GET'), onSuccess, onErrors);*/

export const login = (userName, password, onSuccess, onErrors, reauthenticationCallback) =>
    appFetch('/patients/login', config('POST', {userName, password}),
        authenticatedPatient => {
            setServiceToken(authenticatedPatient.serviceToken);
            setReauthenticationCallback(reauthenticationCallback);
            onSuccess(authenticatedPatient);
        }, 
        onErrors);

export const tryLoginFromServiceToken = (onSuccess, reauthenticationCallback) => {

    const serviceToken = getServiceToken();

    if (!serviceToken) {
        onSuccess();
        return;
    }

    setReauthenticationCallback(reauthenticationCallback);

    appFetch('/patients/loginFromServiceToken', config('POST'),
        authenticatedPatient => onSuccess(authenticatedPatient),
        () => removeServiceToken()
    );

}

export const signUpPatient = (patient, idDietician, onSuccess, onErrors) => {
    console.log("PACIENTE DEL BACK");
    console.log(patient);
    console.log(idDietician);
    appFetch('/patients/signUpPatient', config('POST', {patient, idDietician}), 
        onSuccess,
        onErrors);
    }


export const logout = () => removeServiceToken();

export const updateProfile = (patient, onSuccess, onErrors) =>
    appFetch(`/patients/${patient.id}`, config('PUT', patient),
        onSuccess, onErrors);

export const changePassword = (id, oldPassword, newPassword, onSuccess,
    onErrors) =>
    appFetch(`/patients/${id}/changePassword`, 
        config('POST', {oldPassword, newPassword}),
        onSuccess, onErrors);

File: backend/src/main/java/es/udc/nutritfg/backend/model/services/PatientService.java
package es.udc.nutritfg.backend.model.services;

import es.udc.nutritfg.backend.model.common.exceptions.DuplicateInstanceException;
import es.udc.nutritfg.backend.model.common.exceptions.InstanceNotFoundException;
import es.udc.nutritfg.backend.model.entities.GeneralData;
import es.udc.nutritfg.backend.model.entities.HabitData;
import es.udc.nutritfg.backend.model.entities.Macronutrient;
import es.udc.nutritfg.backend.model.entities.MedicalData;
import es.udc.nutritfg.backend.model.entities.Patient;
import es.udc.nutritfg.backend.model.services.exceptions.IncorrectLoginException;
import es.udc.nutritfg.backend.model.services.exceptions.IncorrectPasswordException;
import es.udc.nutritfg.backend.rest.dtos.AlcoholicDrinkerDto;
import es.udc.nutritfg.backend.rest.dtos.BowelFunctionDto;
import es.udc.nutritfg.backend.rest.dtos.CountryDto;
import es.udc.nutritfg.backend.rest.dtos.GenreDto;
import es.udc.nutritfg.backend.rest.dtos.MaritalStatusDto;
import es.udc.nutritfg.backend.rest.dtos.PhysicalActivityDto;
import es.udc.nutritfg.backend.rest.dtos.SleepQualityDto;
import es.udc.nutritfg.backend.rest.dtos.SmokerDto;

public interface PatientService {

	void signUp(Patient patient, Long idDietician) throws DuplicateInstanceException;

	Patient login(String userName, String password) throws IncorrectLoginException;

	Patient loginFromId(Long id) throws InstanceNotFoundException;

	void changePassword(Long id, String oldPassword, String newPassword)
			throws InstanceNotFoundException, IncorrectPasswordException;

	Patient updateProfile(Long id, String email, GeneralData generalData, HabitData habitData,
			MedicalData medicalData) throws InstanceNotFoundException;


	GenreDto getGenres();

	CountryDto getCountries();

	MaritalStatusDto getMaritalStatus();

	BowelFunctionDto getBowelFunction();

	SleepQualityDto getSleepQuality();

	PhysicalActivityDto getPhysicalActivity();

	SmokerDto getSmoker();

	AlcoholicDrinkerDto getAlcoholicDrinker();

	Patient findPatientById(Long id) throws InstanceNotFoundException;
	
	Patient addMacroDataToPatient(Long id, Macronutrient macro) throws InstanceNotFoundException;
}
